bin/upgrade.py

```diff
@@ -0,0 +1,248 @@
+import problem
+from util import *
+
+import shutil
+
+
+def upgrade_data(p: problem.Problem, bar: ProgressBar) -> None:
+    rename = [
+        ("data/invalid_inputs", "data/invalid_input"),
+        ("data/invalid_answers", "data/invalid_answer"),
+        ("data/invalid_outputs", "data/invalid_output"),
+        ("data/valid_outputs", "data/valid_output"),
+    ]
+    for old_name, new_name in rename:
+        old_path = p.path / old_name
+        new_path = p.path / new_name
+        if old_path.is_dir():
+            if new_path.exists():
+                bar.error(f"can't rename '{old_name}', '{new_name}' already exists", resume=True)
+                continue
+            bar.log(f"renaming '{old_name}' to '{new_name}'")
+            old_path.rename(new_path)
+
+
+def upgrade_generators_yaml(p: problem.Problem, bar: ProgressBar) -> None:
+    generators_yaml = p.path / "generators" / "generators.yaml"
+    if not generators_yaml.is_file():
+        return
+
+    data = read_yaml(generators_yaml)
+    if data is None or "data" not in data:
+        return
+    data = data["data"]
+
+    rename = [
+        ("invalid_inputs", "invalid_input"),
+        ("invalid_answers", "invalid_answer"),
+        ("invalid_outputs", "invalid_output"),
+        ("valid_outputs", "valid_output"),
+    ]
+    for old_name, new_name in rename:
+        if old_name in data:
+            if new_name in data:
+                bar.error(
+                    f"can't rename 'data.{old_name}', 'data.{new_name}' already exists in generators.yaml",
+                    resume=True,
+                )
+                continue
+            bar.log(f"renaming 'data.{old_name}' to 'data.{new_name}' in generators.yaml")
+            data[new_name] = data[old_name]
+            data.pop(old_name)
+
+    # TODO fix testdata.yaml here too?
```

Comment:
Yes, we should move `validator_flags` in `problem.yaml` to `output_validator_args` in `testdata.yaml` in the root of `generators.yaml` :slightly_smiling_face: 

Also, if the `generators.yaml` already contains `testdata.yaml` at any level (not only in root), we should probably migrate `*_validator_flags` to `*_validator_args`?